pkg/runtime/wfengine/activity.go

```diff
@@ -38,14 +44,72 @@
 	a.actorRuntime = actorsRuntime
 }
 
-// InvokeMethod implements actors.InternalActor
-func (*activityActor) InvokeMethod(ctx context.Context, actorID string, methodName string, data []byte) (any, error) {
-	return nil, errors.New("TODO")
+// InvokeMethod implements actors.InternalActor and schedules the background execution of a workflow activity.
+// Activities are scheduled by workflows and can execute for arbitrary lengths of time. Instead of executing
+// activity logic directly, InvokeMethod creates a reminder that executes the activity logic. InvokeMethod
+// returns immediately after creating the reminder, enabling the workflow to continue processing other events
+// in parallel.
+func (a *activityActor) InvokeMethod(ctx context.Context, actorID string, methodName string, data []byte) (any, error) {
+	// This actor doesn't define any methods, and uses the method name and payload as the reminder name and payload.
+	err := a.createReliableReminder(ctx, actorID, methodName, data)
+	return nil, err
 }
 
-// InvokeReminder implements actors.InternalActor
-func (*activityActor) InvokeReminder(ctx context.Context, actorID string, reminderName string, params []byte) error {
-	return errors.New("TODO")
+// InvokeReminder implements actors.InternalActor and executes the activity logic.
+// TODO: Need to review how errors for reminders are handled.
+func (a *activityActor) InvokeReminder(ctx context.Context, actorID string, reminderName string, data any, dueTime string, period string) error {
+	taskEvent, err := backend.UnmarshalHistoryEvent(data.([]byte))
+	if err != nil {
+		return err
+	}
+
+	endIndex := strings.LastIndex(actorID, "#")
+	if endIndex < 0 {
+		return fmt.Errorf("invalid activity actor ID: %s", actorID)
+	}
+	workflowID := actorID[0:endIndex]
+
+	wi := &backend.ActivityWorkItem{
+		SequenceNumber: int64(taskEvent.EventId),
+		InstanceID:     api.InstanceID(workflowID),
+		NewEvent:       taskEvent,
+		Properties:     make(map[string]interface{}),
+	}
+
+	// Executing activity code is a one-way operation. We must wait for the app code to report its completion, which
+	// will trigger this callback channel.
+	// TODO: Need to come up with a design for timeouts. Some activities may need to run for hours but we also need
```

Comment:
Iin addition to the potential for blocking indefinitely if the app crashes (I believe, right? Since the channel will never receive anything and the context won't get cancelled), will this also possibly cause double-activations if daprd crashes and this gets re-invoked? (Or do we just assume the app will handle it even if that does happen?)

Reply:
Yeah, if daprd crashes before the completion callback is received, the activity may be invoked a second time. Since the intent is to provide at-least-once guarantees for activities, we put the burden on the app to deal with this. We could improve it a little by introducing some kind of heartbeat from the app to tell the recovering daprd sidecar that it's still working on that activity, but it's probably not worth the added design complexity, IMO. I don't expect users will run into daprd crashes very often.